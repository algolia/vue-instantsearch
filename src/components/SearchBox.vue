<template>
  <div
    v-if="state"
    :class="suit()"
  >
    <slot
      :current-refinement="currentRefinement"
      :is-search-stalled="state.isSearchStalled"
      :refine="state.refine"
    >
      <search-input
        @focus="$emit('focus', $event)"
        @blur="$emit('blur', $event)"
        @reset="$emit('reset')"
        :placeholder="placeholder"
        :autofocus="autofocus"
        :show-loading-indicator="showLoadingIndicator"
        :should-show-loading-indicator="state.isSearchStalled"
        :submit-title="submitTitle"
        :reset-title="resetTitle"
        :class-names="classNames"
        v-model="currentRefinement"
      >
        <slot
          name="loading-indicator"
          slot="loading-indicator"
        />
        <slot
          name="submit-icon"
          slot="submit-icon"
        />
        <slot
          name="reset-icon"
          slot="reset-icon"
        />
      </search-input>
    </slot>
  </div>
</template>

<script>
import { connectSearchBox } from 'instantsearch.js/es/connectors';
import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';
import SearchInput from './SearchInput.vue';

export default {
  name: 'AisSearchBox',
  mixins: [
    createWidgetMixin({ connector: connectSearchBox }),
    createSuitMixin({ name: 'SearchBox' }),
  ],
  components: {
    SearchInput,
  },
  props: {
    placeholder: {
      type: String,
      default: 'Search hereâ€¦',
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    showLoadingIndicator: {
      type: Boolean,
      default: false,
    },
    submitTitle: {
      type: String,
      default: 'Search',
    },
    resetTitle: {
      type: String,
      default: 'Clear',
    },
    value: {
      type: String,
      default: undefined,
    },
  },
  data() {
    return {
      localValue: '',
    };
  },
  methods: {
    onFormSubmit() {
      const input = this.$el.querySelector('input[type=search]');
      input.blur();
    },
    onFormReset() {
      this.state.refine('');
    },
  },
  computed: {
    isControlled() {
      return typeof this.value !== 'undefined';
    },
    currentRefinement: {
      get() {
        // if the input is controlled, but not up to date
        // this means it didn't search, and we should pretend it was `set`
        if (this.isControlled && this.value !== this.localValue) {
          // eslint-disable-next-line vue/no-side-effects-in-computed-properties
          this.localValue = this.value;
          this.$emit('input', this.value);
          this.state.refine(this.value);
        }
        return this.value || this.state.query || '';
      },
      set(val) {
        this.localValue = val;
        this.state.refine(val);
        if (this.isControlled) {
          this.$emit('input', val);
        }
      },
    },
  },
};
</script>
