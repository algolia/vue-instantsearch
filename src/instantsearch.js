export { createSuitMixin } from './mixins/suit';
export { createWidgetMixin } from './mixins/widget';
export * from './widgets';
export { plugin as default } from './plugin';
export { createInstantSearch } from './util/createInstantSearch';
