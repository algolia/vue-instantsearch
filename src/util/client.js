// TODO: increment this when versioning
const version = 'v2.0.0-alpha.2';

export const addAlgoliaAgent = client =>
  client.addAlgoliaAgent &&
  client.addAlgoliaAgent(`Vue InstantSearch (${version})`);
