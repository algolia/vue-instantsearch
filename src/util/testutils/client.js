export const fakeClient = {
  search: jest.fn(requests =>
    Promise.resolve({
      results: requests.map(({ params: { query } }) => ({ query })),
    })
  ),
};
