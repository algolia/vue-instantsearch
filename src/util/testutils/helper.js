export const createSerializedState = () => ({
  lastResults: {
    _rawResults: [
      {
        hits: [
          {
            objectID: 'doggos',
            name: 'the dog',
          },
        ],
        nbHits: 1071,
        page: 0,
        nbPages: 200,
        hitsPerPage: 5,
        processingTimeMS: 3,
        facets: {
          genre: {
            Comedy: 1071,
            Drama: 290,
            Romance: 202,
          },
        },
        exhaustiveFacetsCount: true,
        exhaustiveNbHits: true,
        query: 'hi',
        queryAfterRemoval: 'hi',
        params:
          'query=hi&hitsPerPage=5&page=0&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&facets=%5B%22genre%22%5D&tagFilters=&facetFilters=%5B%5B%22genre%3AComedy%22%5D%5D',
        index: 'movies',
      },
      {
        hits: [{ objectID: 'doggos' }],
        nbHits: 5131,
        page: 0,
        nbPages: 1000,
        hitsPerPage: 1,
        processingTimeMS: 7,
        facets: {
          genre: {
            Comedy: 1071,
            Drama: 1642,
            Romance: 474,
          },
        },
        exhaustiveFacetsCount: true,
        exhaustiveNbHits: true,
        query: 'hi',
        queryAfterRemoval: 'hi',
        params:
          'query=hi&hitsPerPage=1&page=0&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&attributesToRetrieve=%5B%5D&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=genre',
        index: 'movies',
      },
    ],
    query: 'hi',
    hits: [
      {
        objectID: 'doggos',
        name: 'the dog',
      },
    ],
    index: 'movies',
    hitsPerPage: 5,
    nbHits: 1071,
    nbPages: 200,
    page: 0,
    processingTimeMS: 10,
    exhaustiveFacetsCount: true,
    exhaustiveNbHits: true,
    disjunctiveFacets: [
      {
        name: 'genre',
        data: {
          Comedy: 1071,
          Drama: 1642,
          Romance: 474,
        },
        exhaustive: true,
      },
    ],
    hierarchicalFacets: [],
    facets: [],
    _state: {
      index: 'movies',
      query: 'hi',
      facets: [],
      disjunctiveFacets: ['genre'],
      hierarchicalFacets: [],
      facetsRefinements: {},
      facetsExcludes: {},
      disjunctiveFacetsRefinements: { genre: ['Comedy'] },
      numericRefinements: {},
      tagRefinements: [],
      hierarchicalFacetsRefinements: {},
      hitsPerPage: 5,
      page: 0,
      highlightPreTag: '__ais-highlight__',
      highlightPostTag: '__/ais-highlight__',
    },
  },
});
