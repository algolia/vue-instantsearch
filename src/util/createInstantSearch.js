import instantsearch from 'instantsearch.js/es';
import algoliaHelper from 'algoliasearch-helper';
const { SearchParameters, SearchResults } = algoliaHelper;
import { _objectSpread } from './polyfills';
import { warn } from './warn';

export const createInstantSearch = instantSearchOptions => {
  const search = instantsearch(instantSearchOptions);
  const { searchClient, indexName } = instantSearchOptions;

  // main API for SSR, called in asyncData of a root component which contains instantsearch
  search.findResultsState = params => {
    search.helper = algoliaHelper(
      searchClient,
      indexName,
      _objectSpread({}, params, {
        // parameters set by default
        highlightPreTag: '__ais-highlight__',
        highlightPostTag: '__/ais-highlight__',
      })
    );

    return search.helper.searchOnce().then(({ content: lastResults }) => {
      // The search instance needs to act as if this was a regular `search`
      // but return a promise, since that is the interface of `asyncData`
      search.helper.lastResults = lastResults;
    });
  };

  // make sure correct data is available in each widget's state
  // called in widget mixin
  search.__forceRender = widget => {
    if (!search.helper) {
      warn(
        '`instantsearch.findResultsState()` needs to be called when using `ais-instant-search-ssr`.'
      );
      return;
    }

    widget.init({
      state: search.helper.lastResults._state,
      helper: search.helper,
      templatesConfig: {},
      createURL: () => '#',
      onHistoryChange: () => {},
      instantSearchInstance: search,
    });

    widget.render({
      state: search.helper.lastResults._state,
      results: search.helper.lastResults,
      helper: search.helper,
      templatesConfig: {},
      // TODO: use memory or real router
      createURL: () => '#',
      instantSearchInstance: search,
      searchMetadata: {
        isSearchStalled: false,
      },
    });
  };

  search.getState = () => {
    if (search.helper === null || !search.helper.lastResults) {
      return undefined;
    }
    return {
      lastResults: JSON.parse(JSON.stringify(search.helper.lastResults)),
    };
  };

  // called before app mounts on client
  search.hydrate = instantSearchState => {
    if (!instantSearchState || !instantSearchState.lastResults) {
      warn('The result of `getState()` needs to be passed to `hydrate()`.');
      return;
    }
    const { lastResults } = instantSearchState;
    search.searchParameters = lastResults._state;
    search.helper = algoliaHelper(searchClient, indexName, lastResults._state);
    search.helper.lastResults = new SearchResults(
      new SearchParameters(lastResults._state),
      lastResults._rawResults
    );

    search.hydrated = true;
  };

  // put this in the user's root Vue instance
  // we can then reuse that InstantSearch instance seamlessly from `ais-ssr`
  const rootMixin = {
    provide() {
      return {
        // should be possible to configure this with {camelcase: ['error', {allow: ['^\\$_']}]}
        // but that didn't work
        // eslint-disable-next-line camelcase
        $_ais: search,
      };
    },
  };

  return {
    instantsearch: search,
    rootMixin,
  };
};
