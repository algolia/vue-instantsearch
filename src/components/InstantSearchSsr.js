import { createInstantSearchComponent } from '../util/createInstantSearchComponent';

export default createInstantSearchComponent({
  name: 'AisInstantSearchSsr',
  inject: {
    // should be possible to configure this with {camelcase: ['error', {allow: ['^\\$_']}]}
    // but that didn't work
    // eslint-disable-next-line camelcase
    $_ais: {
      default() {
        throw new Error('`rootMixin` is required when using SSR.');
      },
    },
  },
  data() {
    return {
      instantSearchInstance: this.$_ais,
    };
  },
  render(createElement) {
    return createElement(
      'div',
      {
        class: {
          [this.suit()]: true,
          [this.suit('', 'ssr')]: true,
        },
      },
      this.$slots.default
    );
  },
});
