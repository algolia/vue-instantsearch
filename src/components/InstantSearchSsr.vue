<template>
  <!-- AisSsr is an empty component that will hold other widgets -->
  <div :class="[suit(), suit('', 'ssr')]"><slot /></div>
</template>

<script>
import { createInstantSearch } from '../util/createInstantSearch';

export default createInstantSearch({ ssr: true });
</script>
