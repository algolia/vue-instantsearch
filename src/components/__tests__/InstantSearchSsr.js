import { mount } from '@vue/test-utils';
import instantsearch from 'instantsearch.js/es';
import Vue from 'vue';
import InstantSearchSsr from '../InstantSearchSsr';
import SearchBox from '../SearchBox.vue';
import { createSerializedState } from '../../util/testutils/helper';
import { createInstantSearch } from '../../util/createInstantSearch';
import { fakeClient } from '../../util/testutils/client';

jest.unmock('instantsearch.js/es');

it('requires an injected instantsearch instance ($_ais)', () => {
  expect(() => mount(InstantSearchSsr)).toThrowErrorMatchingInlineSnapshot(
    `"\`rootMixin\` is required when using SSR."`
  );
});

it('renders correctly (empty)', () => {
  const wrapper = mount(InstantSearchSsr, {
    provide: () => ({
      // eslint-disable-next-line camelcase
      $_ais: instantsearch({
        indexName: 'bla',
        searchClient: fakeClient,
      }),
    }),
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly (with slot used)', () => {
  const wrapper = mount(InstantSearchSsr, {
    provide: () => ({
      // eslint-disable-next-line camelcase
      $_ais: instantsearch({
        indexName: 'bla',
        searchClient: fakeClient,
      }),
    }),
    slots: {
      default: '<div>Hi there, this is the main slot</div>',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('does not call __forceRender on second start', () => {
  const { instantsearch: instance, rootMixin } = createInstantSearch({
    indexName: 'bla',
    searchClient: fakeClient,
  });
  instance.hydrate(createSerializedState());
  const forceRenderSpy = jest.spyOn(instance, '__forceRender');

  const wrapper = mount(InstantSearchSsr, {
    ...rootMixin,
    components: {
      AisSearchBox: SearchBox,
    },
    slots: {
      default: SearchBox,
    },
  });

  // called once, for the `SearchBox` widget
  expect(forceRenderSpy).toHaveBeenCalledTimes(1);

  wrapper.destroy();

  mount(InstantSearchSsr, {
    ...rootMixin,
    components: {
      AisSearchBox: SearchBox,
    },
    slots: {
      default: SearchBox,
    },
  });

  // does not call again, since we remove `hydrated` flag on unmount
  expect(forceRenderSpy).toHaveBeenCalledTimes(1);
});

it('does not start too many times', async () => {
  const { instantsearch: instance, rootMixin } = createInstantSearch({
    indexName: 'bla',
    searchClient: fakeClient,
  });

  const startSpy = jest.spyOn(instance, 'start');

  mount(InstantSearchSsr, {
    ...rootMixin,
    components: {
      AisSearchBox: SearchBox,
    },
    slots: {
      default: SearchBox,
    },
  });

  // not started yet (next tick)
  expect(startSpy).toHaveBeenCalledTimes(0);

  mount(InstantSearchSsr, {
    ...rootMixin,
    components: {
      AisSearchBox: SearchBox,
    },
    slots: {
      default: SearchBox,
    },
  });

  // does not yet call again, since same instance needs to unmount first
  expect(startSpy).toHaveBeenCalledTimes(0);

  await Vue.nextTick();

  expect(startSpy).toHaveBeenCalledTimes(1);

  await Vue.nextTick();

  expect(startSpy).toHaveBeenCalledTimes(1);
});
