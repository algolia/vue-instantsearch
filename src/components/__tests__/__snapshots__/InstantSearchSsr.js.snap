// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (empty) 1`] = `

<div class="ais-InstantSearch ais-InstantSearch--ssr">
</div>

`;

exports[`renders correctly (with slot used) 1`] = `

<div class="ais-InstantSearch ais-InstantSearch--ssr">
  <div>
    Hi there, this is the main slot
  </div>
</div>

`;
